[Yong Kai Qi] iP

flairekq opened the pull request:
# Ruby

> "If I had rubies, riches, and crowns, I'd buy the whole world and change things around." - Bob Dylan ([source](https://www.azquotes.com/quote/987912)) 

Ruby frees your mind from having to remember things you need to do. It's,
- text-based
- ~~easy~~ very easy to learn

All you need to do is,

1. download it from [here](https://github.com/flairekq/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too! Here's a small snippet of the `run` code: 
```java
public void run() {
      this.ui.showWelcome();
      boolean isRun = true;
      while (isRun) {
          try {
              String command = this.ui.readCommand();
              Command c = Parser.parse(command, this.tasks);
              c.execute(this.tasks, this.ui, this.storage);
              isRun = c.isRunProgram();
          } catch (DukeException e) {
              this.ui.showError(e.getMessage());
          }
      }
}
```

Commit:
Add Gradle support

flairekq committed:
Update Duke.java

Update Duke.java to greet, echo and exit when the user types bye.

flairekq committed:
Update Duke.java

Update Duke.java to add the texts entered by the user and display them back to the user when requested (list).

flairekq committed:
Add Task.java and update Duke.java

Add Task.java to represent tasks.
Update Duke.java to have the ability to mark tasks as done and not done.

flairekq committed:
Add Deadline.java, Event.java, Todo.java and update Duke.java

Update Duke.java to support the tracking of 3 types of tasks: ToDos, Deadlines & Events.
Add Deadline.java, Event.java, Todo.java to use inheritance to support multiple task types

flairekq committed:
Update Duke.java, EXPECTED.TXT and input.txt

Update Duke.java - Removed additional spacing at the printMsg of addTask()
Update input.txt and EXPECTED.TXT to test with the latest features of Duke.java

flairekq committed:
Update Duke.java

Update Duke.java to handle input errors

flairekq committed:
Update Duke.java

Update Duke.java to delete tasks from list of tasks

flairekq committed:
Add DukeException.java and update Duke.java

Add DukeException.java to have a custom Exception class to handle the invalid inputs.
Update Duke.java to utilize DukeException

flairekq committed:
Update bot to have save and load feature

- Loads the data from the hard disk ("data/duke.txt") when the bot starts up.
- Update the hard disk ("data/duke.txt") whenever the task list changes

flairekq committed:
Update bot to understand dates

Deadline class now stores the property by as java.time.LocalDate. It prints out the date stored in the format of MMM dd yyyy

flairekq committed:
Merge branch 'branch-Level-7'

flairekq committed:
Merge branch 'branch-Level-8'

flairekq committed:
Update the save format of Deadline

Update the save format of Deadline to print out the property by (LocalDate) in MMM dd yyyy format

flairekq committed:
Make the code more OOP and update the save format of Deadline

Extracted the following classes:
Ui: deals with interactions with the user
Storage: deals with loading tasks from the file and saving tasks in the file
Parser: deals with making sense of the user command
TaskList: contains the task list and operations to manipulate the list
Command: AddCommand, DeleteCommand, ListCommand, ExitCommand, UpdateCommand that inherit from an abstract Command class to write the main logic of the App

Update the save format of Deadline to remove the formatting of property "by"

flairekq committed:
Organize into Packages

Divide classes into packages
command:  AddCommand, Command, DeleteCommand, ExitCommand, ListCommand, UpdateCommand
duke: Duke, DukeException, Parser, Storage, Ui
task: Deadline, Event, Task, TaskList, Todo

flairekq committed:
Add JUnit tests

Add Junit tests to test the behavior of the following classes' codes:
- Deadline
- Event
- Todo

flairekq committed:
Update Storage to create text file to save tasks if it does not exist

flairekq committed:
Add JavaDoc comments

flairekq committed:
Follow the Coding Standard

flairekq committed:
Update bot to find matching tasks based on the search keyword

flairekq committed:
Merge branch 'branch-A-JavaDoc'

flairekq committed:
Merge branch 'branch-A-CodingStandard'

flairekq committed:
Merge branch 'branch-Level-9'

flairekq committed:
Remove unused import statement

dionegoh commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,116 @@
+     * @param updatedDetails The new save format of the task to be updated.
+     * @throws DukeException If there are issues accessing or writing to the data file .
+     */
+    public void updateSavedTasks(String oldDetails, String updatedDetails) throws DukeException {
```
Perhaps this method could be modified to reduce deep nesting as it currently has more than 3 levels of indentations!

dionegoh commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * The Command class is an abstract class to represent an actual command by the user.
+ */
+public abstract class Command {
```
I like how you have created an abstract Command class with an abstract method `execute` for the different types of commands to inherit from and be executed differently! 👍

dionegoh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Runs the bot application until the user inputs the exit command.
+     */
+    public void run() {
```
I like how clean this piece of code is and how OOP is used to make the main logic of the application easier to comprehend! 😄 Great job!

dionegoh commented on src/main/java/command/AddCommand.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     * Executes the add command and adds the new task to the TaskList and stores it in the Storage as well.
+     * @param tasks TaskList that stores the current list of tasks.
```
Perhaps an empty line could be added between description and parameter section to adhere to the Java Coding standard! Similarly, for other javadocs in the code as well.

dionegoh reviewed (commented):
LGTM! 👍 Just a few nits to fix. Keep up the good work! 😄

flairekq committed:
Merge remote-tracking branch 'origin/add-gradle-support'

flairekq committed:
Use Gradle

Set up gradle to build and run Duke using gradle and also run unit tests

flairekq committed:
Use CheckStyle

Use checkStyle to detect coding style violations and corrected based on it

flairekq committed:
Add GUI

Add a GUI to Ruby using JavaFX technology.

punnyhuimin commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,35 @@
+    protected boolean isRun;
+
+    public Command(String command, String[] tokenizedCommand) {
+        this.command = command;
```
Is it necessary to reference the variable `command` with 'this'?

punnyhuimin commented on src/main/java/command/FindCommand.java:
```diff
@@ -0,0 +1,30 @@
+
+    /**
+     * Executes the find command and displays the list of tasks that matches the keyword.
+     * @param tasks TaskList that stores the current list of tasks.
```
Do you need to put `@return` in a new line? I noticed this in the other files as well. You could consider getting IntelliJ to [autoformat](https://stackoverflow.com/questions/17879475/how-enable-auto-format-code-for-intellij-idea) your files. 😄 
<img width="355" alt="Screenshot 2022-02-03 at 8 38 45 PM" src="https://user-images.githubusercontent.com/61443625/152344331-8a6dab6c-b359-4e53-8f60-1361858906f8.png">

punnyhuimin commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,58 @@
+        new DukeException().validateInputs(command, tokenizedCommand, tasks);
+        Command c = null;
+        switch (tokenizedCommand[0]) {
+        case "bye":
```
Great formatting of the switch case statements 😄

punnyhuimin commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,76 @@
+        this.tasks = new ArrayList<>();
+    }
+
+    public TaskList(ArrayList<Task> tasks) {
```
Maybe you could add JavaDocs here? 😃

punnyhuimin commented on src/main/java/duke/MainWindow.java:
```diff
@@ -0,0 +1,67 @@
+package duke;
+
+import javafx.fxml.FXML;
```
Good effort in importing classes explicitly. 👍

punnyhuimin reviewed (commented):
LGTM overall! Minor changes can be done, but overall it looks great! 💯

flairekq committed:
Fix null error when loading fxml

alfredkohhh reviewed (commented):
Good job on spotting the unused import statement !

alfredkohhh reviewed (commented):
I like that you how you separated the import statements, as opposed to the IDE auto importing the * ! Good job

flairekq committed:
Add linespace to Duke.java

flairekq committed:
Merge branch 'branch-Level-10'

flairekq committed:
Update gradle file and generate jar file

Update gradle file to change the mainClassName to duke.Launcher as it is the class that launches the program.

Generate jar file to have a new jar file that is able to run the program with the latest GUI feature

flairekq committed:
Use assertions

Use assert feature (not JUnit assertions) to document important assumptions that should hold at various points in the code

flairekq committed:
Improve Code Quality

Refactor codes to improve the code quality where necessary.

flairekq committed:
Merge pull request #1 from flairekq/branch-A-Assertions

Use assertions

flairekq committed:
Merge pull request #2 from flairekq/branch-A-CodeQuality

Improve Code Quality

flairekq committed:
Set up CI

Use Github Actions to set up CI.

flairekq committed:
Merge remote-tracking branch 'origin/master'

flairekq committed:
Sort tasks managed by the App

Implement sorting of the tasks by type of tasks where Deadline comes first, followed by Event, followed by Todo. 

For Deadline:
* sort by due date 
* sort by description in ascending order

For Event & Todo: 
* sort by description in ascending order

flairekq committed:
Improve the GUI

Tweak padding, colors and background image of the dialog container to make the GUI more pleasing to look at

flairekq committed:
Fix Storage bug

Fix Storage bug of not adding a new line when storing new task.

flairekq committed:
Add a User Guide

flairekq committed:
Update javadocs comments

flairekq committed:
Fix Storage bug and add exit app

Fix Storage bug of adding inconsistent amount of new lines to the data.txt

Exits app after 3 seconds from user giving a command of "bye"
